nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py

```diff
@@ -134,91 +137,152 @@ def __getitem__(self, idx):
+            text = self.prompt_template.replace(context_string, original_context).replace(label_string, label)
+            if self.query_key is not None:
+                query_string = f'{{{self.query_key}}}'
+                assert query_string in self.prompt_template
+                context = context.replace(query_string, query)
+                text = text.replace(query_string, query)
+
+        elif self.separate_prompt_and_response_with_newline:
+            if self.query_key is not None:
+                context = context + '\n' + query
+            text = context + '\n' + label
+        else:
+            if self.query_key is not None:
+                context = context + ' ' + query
+            text = context + ' ' + label
 
         if self.virtual_tokens:
             # (@adithyare) we are going to insert "pad/eos" tokens in the beginning of the text and context
             # these pad/eos tokens are placeholders for virtual tokens
             pre_pad = [self.tokenizer.eos_id] * self.virtual_tokens
         else:
             pre_pad = []
+
         tokenized_text = pre_pad + self.tokenizer.text_to_ids(text)
         context_ids = pre_pad + self.tokenizer.text_to_ids(context)
         answer_ids = tokenized_text[len(context_ids) :]
-
-        # for the long context cases, collate_fn includes self.tokens_to_generate for padding
-        total_ids = len(context_ids) + max(len(answer_ids), self.tokens_to_generate)
-        if self.add_bos:
-            total_ids += 1
-        if self.add_sep:
-            total_ids += 1
-        if self.add_eos:
-            total_ids += 1
-
-        # If the total number of token is greater than the max, we will try to truncate the answer
-        if total_ids > self.max_seq_length:
-            truncation_length = total_ids - self.max_seq_length
-            if self.truncation_field == "answer":
-                answer_ids = answer_ids[: -min(truncation_length, len(answer_ids))]
-            elif self.truncation_field == "context":
-                context_ids = context_ids[: -min(truncation_length, len(context_ids))]
-
-        if len(context_ids) > self.max_seq_length:
-            context_ids = context_ids[: self.max_seq_length]
-
-        assert len(context_ids) <= self.max_seq_length
         input_ids = context_ids
 
         answer_start_idx = len(input_ids)
         # Adds sep token between text/prompt and answer
         if self.add_sep:
+            context_ids = context_ids + [self.sep_id]
             input_ids = input_ids + [self.sep_id]
             answer_start_idx += 1
 
         input_ids = input_ids + answer_ids
 
         if self.add_bos:
+            context_ids = [self.tokenizer.bos_id] + context_ids
             input_ids = [self.tokenizer.bos_id] + input_ids
             answer_start_idx += 1
-        if self.add_eos:
+
+        # Only training need to consider eos token
+        if self.add_eos and self.tokens_to_generate == 0:
```

Comment:
don't we want to stop generation during inference if eos is omitted?

Reply:
But during generation, we don't need to add EOS right? EOS is generated by model